streamer/src/nonblocking/quic.rs

```diff
@@ -85,6 +88,119 @@ struct PacketAccumulator {
+        let num_extra_updates = time_since_last_update_ms.saturating_sub(1) / update_interval_ms;
+
+        let mut updated_load_ema = Self::ema_function(
+            self.current_load_ema.load(Ordering::Relaxed) as u128,
+            self.load_in_recent_interval.load(Ordering::Relaxed) as u128,
+        );
+
+        for _ in 0..num_extra_updates {
+            updated_load_ema = Self::ema_function(
+                updated_load_ema,
+                self.load_in_recent_interval.load(Ordering::Relaxed) as u128,
+            );
+        }
+
+        self.current_load_ema
+            .store(updated_load_ema as u64, Ordering::Relaxed);
+        self.stats
+            .stream_load_ema
+            .store(updated_load_ema as usize, Ordering::Relaxed);
+    }
+
+    fn update_ema_if_needed(&self) {
+        // Read lock enables multiple connection handlers to run in parallel if interval is not expired
+        if Instant::now().duration_since(*self.last_update.read().unwrap())
+            >= STREAM_LOAD_EMA_INTERVAL_MS
+        {
+            let mut last_update_w = self.last_update.write().unwrap();
+            // Recheck as some other thread might have updated the ema since this thread tried to acquire the write lock.
+            let since_last_update = Instant::now().duration_since(*last_update_w);
+            if since_last_update >= STREAM_LOAD_EMA_INTERVAL_MS {
+                *last_update_w = Instant::now();
+                self.update_ema(since_last_update.as_millis());
+            }
+        }
+    }
+
+    fn increment_load(&self) {
+        self.load_in_recent_interval.fetch_add(1, Ordering::Relaxed);
+        self.update_ema_if_needed();
+    }
+    fn decrement_load(&self) {
+        self.load_in_recent_interval.fetch_sub(1, Ordering::Relaxed);
+        self.update_ema_if_needed();
+    }
+
+    fn available_load_capacity_in_duration(
+        &self,
+        stake: u64,
+        total_stake: u64,
+        duration_ms: u128,
+    ) -> u128 {
+        let ema_window_ms =
+            STREAM_LOAD_EMA_INTERVAL_MS.as_millis() * STREAM_LOAD_EMA_INTERVAL_COUNT;
+        let max_load_in_ema_window: u128 = (MAX_STREAMS_PER_MS as u128
+            - Percentage::from(MAX_UNSTAKED_STREAMS_PERCENT).apply_to(MAX_STREAMS_PER_MS as u128))
+            * ema_window_ms;
+
+        // If the current load is low, cap it to 25% of max_load.
+        let current_load = cmp::max(
+            self.current_load_ema.load(Ordering::Relaxed) as u128,
+            max_load_in_ema_window / 4,
+        );
+
+        // Formula is (max_load ^ 2 / current_load) * (stake / total_stake)
+        // To avoid overflow due to multiplication, we are doing the following
+        // (max_load / current_load) * (max_load / total_stake) * stake
+        let capacity_in_ema_window = ((max_load_in_ema_window as f64 / current_load as f64)
+            * (max_load_in_ema_window as f64 / total_stake as f64)
+            * stake as f64) as u128;
```

Comment:
main motivation for retyping the constants was to _not_ do this math in floating point (total stake already loses precision as an `f64`). only this math needs to be done in `u128` the return can remain a `u64`

Reply:
https://github.com/solana-labs/solana/pull/34586/files#diff-4540f23a8931e5cbcbfb8a2d2ba5bcf8c20c466fadb5c770190de839a846847dR119 is the math that benefited from using u128.

I tried doing similar changes to the math here. It can potentially cause overflows, since `total_stake` is in lamports. So we need to multiply `max_load_in_ema_window` with a large enough number that ensures `max_load_in_ema_window / total_stake` is not 0. But in a weird configuration, total_stake can be low (which is unlikely, but the code will support it), and cause overflows.

Reply:
How does this look as a replacement?

```
        let capacity_in_ema_window =
            (max_load_in_ema_window * max_load_in_ema_window * stake as u128)
                / (current_load * total_stake as u128);
```